server/api.go

```diff
@@ -79,6 +79,38 @@ func (d dexAPI) CreateClient(ctx context.Context, req *api.CreateClientReq) (*ap
 	}, nil
 }
 
+func (d dexAPI) UpdateClient(ctx context.Context, req *api.UpdateClientReq) (*api.UpdateClientResp, error) {
+	if req.Id == "" {
+		return nil, errors.New("update client: no client ID supplied")
+	}
+
+	err := d.s.UpdateClient(req.Id, func(old storage.Client) (storage.Client, error) {
+		if req.RedirectUris != nil && len(req.RedirectUris) > 0 {
+			old.RedirectURIs = req.RedirectUris
+		}
+		if req.TrustedPeers != nil && len(req.TrustedPeers) > 0 {
+			old.TrustedPeers = req.TrustedPeers
+		}
+		old.Public = req.Public
```

Comment:
So, we're "patchy" for `Name` and `LogoUrl` -- empty-string means no update -- but not for `Public`? There is ways to do this in GRPC, so I wonder if we want to do that.

Concretely, if `Public` was of type [`BoolValue`](https://godoc.org/github.com/golang/protobuf/ptypes/wrappers#BoolValue), we could differentiate "not provided" from "false". I suppose if we do that, we should do the same with the other fields; as you currently cannot
- unset trusted peers
- unset redirect URIs
- set the name to empty-string
- set the logoURL to empty-string

💭 I guess one could argue whether this is needed. But as it stands, you cannot update the client without _setting its public value_; that at least should be fixed IMHO

Reply:
What would be your suggestion in this case? if it is a private client and the user tries to update to public we should return an error?

Reply:
It seems that I need to use an external package in the proto file in order to get these types working. 
```
import "google/protobuf/wrappers.proto";
```
Some details here: https://github.com/golang/protobuf/issues/438

I can't get it working. If you have any suggestion how to make this working, I am up to it. 

An alternative solution is to not update the `Public` field in the update operation.

Reply:
I tried to address your concerns in the last commit. Please let me know if this solution is fine for you.

Reply:
Yeah, I'm not sure how to best deal with this one. An alternative to introducing wrappers could be to use a [FieldMask](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#fieldmask) in UpdateClientReq. And of course, we could just leave it as-is. What do you think, @ericchiang?

Re 49fa5ee6e8271d433cd9a3b00c3533495e285c3e, I'm not sure I follow... From what I can tell, it's now an append-only operation? I.e. whatever is in the UpdateClientReq.RedirectUris will be _added_ to old.RedirectUris; but nothing could ever get removed? I'm not sure if this really is what we want... I think something like
```go
		if req.RedirectUris != nil {
			old.RedirectURIs = req.RedirectUris
		}
```
would do the trick -- it lets you determine what the RedirectUris are _after the update_, and it allows you to set them to `[]string{}`. Before, what last thing wasn't possible (unless I'm mistaken). Likewise for TrustedClients.

Thanks for bearing with me here 😃

Reply:
😄 This was what I understood from your first comment. I reverted the last commit, and also allow to update the redirect URIs and trusted peers to []string{}. 

The `Public` filed update was removed as well. I think is not required to be updated after the initial registration. 

Hope this is what you are expecting! Thanks